runtime/jailer.go

```diff
@@ -0,0 +1,299 @@
+			return err
+		}
+	} else {
+		return fmt.Errorf("unsupported mode: %v", stat.Mode)
+	}
+
+	if err := os.Chmod(dst, 0666); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// exposeBlockDeviceToJail will call mknod on the block device to ensure
+// visibility of the device
+func exposeBlockDeviceToJail(dst string, rdev int) error {
+	if err := syscall.Mknod(dst, syscall.S_IFBLK, int(rdev)); err != nil {
+		return err
+	}
+	return nil
+}
+
+type jailer struct {
+	jailPath string
+}
+
+func newJailer(rootpath string) jailer {
+	return jailer{
+		jailPath: rootpath,
+	}
+}
+
+// JailPath returns the base directory from where the jail binary will be ran
+// from
+func (j jailer) JailPath() string {
+	return j.jailPath
+}
+
+// RootPath returns the root fs of the jailed system.
+func (j jailer) RootPath() string {
+	return filepath.Join(j.jailPath, "rootfs")
+}
+
+func (j *jailer) BuildHandler(logger *logrus.Entry, cfg *Config, socketPath *string, vmID string) firecracker.Handler {
+	jailPath := j.JailPath()
+	rootPath := j.RootPath()
+	*socketPath = filepath.Join(rootPath, "api.socket")
+
+	return firecracker.Handler{
+		Name: jailerHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			// Create the proper paths needed for the runc jailer
+			logger.Debugf("Creating root drive path %v", rootPath)
+			if err := os.MkdirAll(rootPath, 0777); err != nil {
+				return errors.Wrapf(err, "failed to create root path: %v", rootPath)
+			}
+
+			binPath := filepath.Join(rootPath, "usr", "local", "bin")
+			logger.Debugf("Creating /usr/local/bin %v", binPath)
+			if err := os.MkdirAll(binPath, 0777); err != nil {
+				return errors.Wrapf(err, "failed to create /usr/local/bin in root path: %v", binPath)
+			}
+
+			devPath := filepath.Join(rootPath, "dev")
+			logger.Debugf("Creating /dev/net %v", devPath)
+			if err := os.MkdirAll(filepath.Join(devPath, "net"), 0777); err != nil {
+				return errors.Wrapf(err, "failed to create device path: %v", devPath)
+			}
+
+			contentsPath := filepath.Join(rootPath, "var", "lib", "firecracker-containerd")
+			logger.Debugf("Creating firecracker contents path %v", contentsPath)
+			if err := os.MkdirAll(contentsPath, 0777); err != nil {
+				return errors.Wrapf(err, "failed to create contents path: %v", contentsPath)
+			}
+
+			rootPathToConfig := filepath.Join(jailPath, "config.json")
+			logger.Debugf("Copying config to %v", rootPathToConfig)
+			if err := copyFile(runcConfigPath, rootPathToConfig); err != nil {
+				return errors.Wrapf(err, "failed to copy config from %v to %v", runcConfigPath, rootPathToConfig)
+			}
+
+			logger.Debugf("Overwritting process args of config")
+			if err := overwriteConfig(cfg, filepath.Base(m.Cfg.SocketPath), rootPathToConfig); err != nil {
+				return errors.Wrap(err, "failed to overwrite config.json")
+			}
+
+			logger.Debugf("chdir to %v", jailPath)
+			if err := os.Chdir(jailPath); err != nil {
```

Comment:
Do we execute this chdir after fork()? chdir is having process-wide effect. It is better to not depend on the working directory.

Reply:
This must occur since we are using `runc` as a jailer. I do not know of another way of doing this. So if you have any answers let me know :p